Added SoftBloomFilter

codex128 opened the pull request:
The filter introduced in this PR produces much cleaner and visually natural results than BloomFilter. Only supports the equivalent of GlowMode.Scene (no select object glow).

Basic usage is similar to BloomFilter:
```java
FilterPostProcessor fpp = new FilterPostProcessor(assetManager);
PBRBloomFilter bloom = new PBRBloomFilter();
fpp.addFilter(bloom);
viewPort.addProcessor(fpp);
```

Note: name changed from PBRBloomFilter to SoftBloomFilter.

capdevon commented on jme3-effects/src/main/java/com/jme3/post/filters/PBRBloomFilter.java:
```diff
@@ -0,0 +1,262 @@
+    /**
+     * Gets the number of downsampling/upsampling passes per step.
+     * 
+     * @return glow factor
```
"`@return glow factor`": error in the javadoc
should be "`@return number of downsampling/upsampling passes`"

capdevon commented on jme3-effects/src/main/java/com/jme3/post/filters/PBRBloomFilter.java:
```diff
@@ -0,0 +1,262 @@
+    public float getGlowFactor() {
+        return glowFactor;
+    }
+    
```
The read/write methods for serializing the filter are missing

eg
```java
    @Override
    public void write(JmeExporter ex) throws IOException {
        super.write(ex);
        OutputCapsule oc = ex.getCapsule(this);
        oc.write(glowMode, "glowMode", GlowMode.Scene);
        oc.write(blurScale, "blurScale", 1.5f);
        oc.write(exposurePower, "exposurePower", 5.0f);
        oc.write(exposureCutOff, "exposureCutOff", 0.0f);
        oc.write(bloomIntensity, "bloomIntensity", 2.0f);
        oc.write(downSamplingFactor, "downSamplingFactor", 1);
    }

    @Override
    public void read(JmeImporter im) throws IOException {
        super.read(im);
        InputCapsule ic = im.getCapsule(this);
        glowMode = ic.readEnum("glowMode", GlowMode.class, GlowMode.Scene);
        blurScale = ic.readFloat("blurScale", 1.5f);
        exposurePower = ic.readFloat("exposurePower", 5.0f);
        exposureCutOff = ic.readFloat("exposureCutOff", 0.0f);
        bloomIntensity = ic.readFloat("bloomIntensity", 2.0f);
        downSamplingFactor = ic.readFloat("downSamplingFactor", 1);
    }
```

capdevon commented on jme3-examples/src/main/java/jme3test/post/TestPBRBloom.java:
```diff
@@ -0,0 +1,148 @@
+
+    }
+
+ 
```
remove extra spaces

capdevon commented on jme3-examples/src/main/java/jme3test/post/TestPBRBloom.java:
```diff
@@ -0,0 +1,148 @@
+        
+        rootNode.addControl(new EnvironmentProbeControl(assetManager, 256));
+        
+        fpp=new FilterPostProcessor(assetManager);
```
Add the spaces before and after =

capdevon commented on jme3-examples/src/main/java/jme3test/post/TestPBRBloom.java:
```diff
@@ -0,0 +1,148 @@
+        rootNode.addControl(new EnvironmentProbeControl(assetManager, 256));
+        
+        fpp=new FilterPostProcessor(assetManager);
+        PBRBloomFilter bloom=new PBRBloomFilter();
```
Add the spaces before and after =

capdevon commented on jme3-examples/src/main/java/jme3test/post/TestPBRBloom.java:
```diff
@@ -0,0 +1,148 @@
+        fpp=new FilterPostProcessor(assetManager);
+        PBRBloomFilter bloom=new PBRBloomFilter();
+        fpp.addFilter(bloom);
+        //ToneMapFilter toneMap = new ToneMapFilter();
```
Remove the ToneMapFilter code if unused

capdevon commented on jme3-effects/src/main/java/com/jme3/post/filters/PBRBloomFilter.java:
```diff
@@ -0,0 +1,262 @@
+     * @see #setNumSamplingPasses(int)
+     */
+    public PBRBloomFilter(int numSamplingPasses) {
+        this.numSamplingPasses = numSamplingPasses;
```
Please add the call to the empty constructor to initialize the filter name.

eg:
```java
public PBRBloomFilter(int numSamplingPasses) {
        this();
        this.numSamplingPasses = numSamplingPasses;
}
```

capdevon commented on jme3-effects/src/main/java/com/jme3/post/filters/PBRBloomFilter.java:
```diff
@@ -0,0 +1,254 @@
+     */
+    public void setNumSamplingPasses(int numSamplingPasses) {
+        if (numSamplingPasses <= 0) {
+            throw new IllegalArgumentException("Expected number of sampling passes to be greater than zero (found: " + numSamplingPasses + ").");
```
Here's a better exception message:
```java
"Number of sampling passes must be greater than zero. Received: " + numSamplingPasses
```

capdevon commented on jme3-effects/src/main/java/com/jme3/post/filters/PBRBloomFilter.java:
```diff
@@ -0,0 +1,254 @@
+     * <p>
+     * default=5
+     * 
+     * @param numSamplingPasses number of passes per donwsampling/upsampling step
```
better param description:
```java
@param numSamplingPasses The number of sampling passes to use. Must be greater than zero.
```

Add this after param description:
```java
@throws IllegalArgumentException if the provided number of sampling passes is less than or equal to zero.
```

eg:
```java
    /**
     * Sets the number of sampling passes for the filter.
     * <p>
     * This method sets the number of times the filter will sample the input texture
     * when rendering. A higher number of sampling passes typically results in a
     * higher quality filtered image, but also comes at a performance cost.
     * 
     * @param numSamplingPasses The number of sampling passes to use. Must be
     *                          greater than zero.
     * @throws IllegalArgumentException if the provided number of sampling passes is
     *                                  less than or equal to zero.
     */
```

capdevon commented on jme3-effects/src/main/java/com/jme3/post/filters/PBRBloomFilter.java:
```diff
@@ -0,0 +1,254 @@
+    private AssetManager assetManager;
+    private RenderManager renderManager;
+    private ViewPort viewPort;
+    private int width, height;
```
Declare "height" on a separate line.

capdevon commented on jme3-effects/src/main/java/com/jme3/post/filters/PBRBloomFilter.java:
```diff
@@ -0,0 +1,254 @@
+    }
+    
+    @Override
+    protected void initFilter(AssetManager am, RenderManager rm, ViewPort vp, int width, int height) {
```
You can simplify the creation of duplicate variables for `w` and `h`. 
Perhaps you can simplify the code in this way:

```java
    @Override
    protected void initFilter(AssetManager am, RenderManager rm, ViewPort vp, int w, int h) {

        this.assetManager = am;
        this.renderManager = rm;
        this.viewPort = vp;
        this.postRenderPasses = new LinkedList<>();
        this.width = w;
        this.height = h;

        Pass[] downsamplingPasses = new Pass[numSamplingPasses];
        Pass[] upsamplingPasses = new Pass[numSamplingPasses];
        
        Renderer renderer = renderManager.getRenderer();

        // the rest of the code is the same...
}
```

capdevon commented:
* Does the PBRBloomFilter work only with PBR materials ? 
* What effect does it have on Lighting materials (the teapot and the box) ? 
* Can you copy/paste a screenshot in the comments please ? 

The java code looks ok to me, but I am not experienced enough to do a review of the shader code. 
So you need approval from a shader expert and a core member for final integration.

codex128 commented:
> Does the PBRBloomFilter work only with PBR materials?

In general, yes, because this filter relies heavily on HDR.

> What effect does it have on Lighting materials (the teapot and the box in the test class)?

Same effect, technically, though not noticeable because Lighting.j3md doesn't utilize HDR. I copied TestBloomFilter.java, so that is why the teapot is in the test.

> Can you copy/paste a screenshot in the comments please?

![](https://github.com/codex128/ImageUpload/blob/master/better-pbrbloom.png?raw=true)

riccardobl commented:
This is a multipass bloom filter that uses hardware filtering to increase the samples, similar to the mipmap bloom filter but it doesn't relies on the opengl mipmap generation function (that is driver dependent and not guaranteed to run on the gpu).

This is much better than the default bloom filter and it has a different optimization (uses hardware filtering instead of two pass blur) so it might be pretty much equivalent or even better from  a performance standpoint. 

My only suggestion is to find a different name, since calling it PBR might be confusing.

I have one question though, what is the default min filter for the filter `getRenderedTexture()` ? This should work best with MinFilter.BilinearNoMipMaps , can you check that? @codex128

codex128 commented:
> My only suggestion is to find a different name, since calling it PBR might be confusing.

Good idea, though PBR Bloom is what the original article calls it. Perhaps SoftBloomFilter?

> what is the default min filter for the filter `getRenderedTexture()` ?

Yup, the default min filter for pass textures is NearestNoMipMaps.

riccardobl commented:
This is a multipass bloom filter that uses hardware sampling, kinda like the mipmap bloom filter-

> > My only suggestion is to find a different name, since calling it PBR might be confusing.
> 
> Good idea, though PBR Bloom is what the original article calls it. Perhaps SoftBloomFilter?

Yes SoftBloomFilter might be a good option.

> > what is the default min filter for the filter `getRenderedTexture()` ?
> 
> Yup, the default min filter for pass textures is NearestNoMipMaps, so that should be good.

Can you try with **Bilinear**NoMipMaps and show the results? This will enable hardware filtering when downscaling , the result should be smoother.
Might be worth trying also Bilinear MagFilter when upscaling.

codex128 commented:
> Can you try with **Bilinear**NoMipMaps and show the results?

Oh, sorry, my bad. :stuck_out_tongue: Here is the result with min=BilinearNoMipMaps and mag=Bilinear:

![](https://github.com/codex128/ImageUpload/blob/master/bilinearNoMipMapsGlow.png?raw=true)

It's difficult to tell, but I think the default min/mag filters produce *slightly* rougher glow. For reference:

![](https://github.com/codex128/ImageUpload/blob/master/better-pbrbloom.png?raw=true)

The differences become more apparent if you zoom in on the glowing sections.

riccardobl commented:
It should be more visible with complex scenes.
One last thing, can you make it on by default but toggleable with `setBilinearFiltering(false)`?

codex128 commented:
I have implemented that. I also added something to suppress the number of downsampling/upsampling passes made so that textures do not have width or height <= 0.

capdevon commented on jme3-effects/src/main/java/com/jme3/post/filters/SoftBloomFilter.java:
```diff
@@ -0,0 +1,335 @@
+        super.write(ex);
+        OutputCapsule oc = ex.getCapsule(this);
+        oc.write(numSamplingPasses, "numSamplingPasses", 5);
+        oc.write(glowFactor, "glowFactor", 0.05f);
```
serialize `bilinearFiltering` field

`oc.write(bilinearFiltering, "bilinearFiltering", true);`

codex128 replied:
Keep forgetting about that one, thanks for pointing it out. :+1:

capdevon commented on jme3-effects/src/main/java/com/jme3/post/filters/SoftBloomFilter.java:
```diff
@@ -0,0 +1,335 @@
+        super.read(im);
+        InputCapsule ic = im.getCapsule(this);
+        numSamplingPasses = ic.readInt("numSamplingPasses", 5);
+        glowFactor = ic.readFloat("glowFactor", 0.05f);
```
deserialize `bilinearFiltering` field

`bilinearFiltering = ic.readBoolean("bilinearFiltering", true);`

riccardobl commented:
> I have implemented that. I also added something to suppress the number of downsampling/upsampling passes made so that textures do not have width or height <= 0.

Just to be sure to not hurt the feeling of any driver, could you please make it <=2 ?

riccardobl reviewed (commented):
Can you rename PBRBloomFinal.frag and PBRBloomFinal.j3md to SoftBloomFinal.frag/j3md?

riccardobl reviewed (approved):
Perfect. Thank you.
